src/v/cluster/controller_backend.cc

```diff
@@ -885,522 +690,491 @@ void controller_backend::housekeeping() {
+                : std::nullopt,
+      claim_it != _ntp_claims.end() ? std::optional{claim_it->second}
+                                    : std::nullopt);
+
+    // Cleanup older revisions that should not exist on this node. This is
+    // typically done after the replicas update is finished.
 
-    auto partition = _partition_manager.local().get(ntp);
-    /*
-     * current change is obsolete, configuration is already updated with
-     * more recent change, do nothing, even if this node is required for the new
-     * revision to catch up we will keep the partition alive until one of the
-     * other nodes will send update_finished command
-     */
     if (
-      partition
-      && partition->group_configuration().revision_id() > command_rev) {
+      claim_it != _ntp_claims.end()
+      && (!expected_log_rev || (claim_it->second.log_revision < *expected_log_rev))) {
         vlog(
           clusterlog.trace,
-          "[{}] found newer revision, finishing reconfiguration to: {}",
+          "[{}] dropping obsolete partition claim {}",
           ntp,
-          data.target_assignment.replicas);
-        co_return std::error_code(errc::success);
+          claim_it->second);
+        // TODO: if hosting = true, delete persistent kvstore state.
+        _ntp_claims.erase(claim_it);
+        claim_it = _ntp_claims.end();
     }
 
-    /**
-     * Check if target assignment has node and core local replicas
-     */
-    if (!has_local_replicas(_self, data.target_assignment.replicas)) {
-        /**
-         * if no replicas are expected on current node/shard and partition
-         * doesn't exists, the update is finished
-         */
-        if (!partition) {
-            co_return std::error_code(errc::success);
-        }
-
-        /**
-         * Cross core partition migration
-         *
-         * Current partition have to be migrated to other shard of the current
-         * node, in this situation we have to
-         * 1) shutdown partition instance
-         * 2) create instance on target remote core
-         */
-        if (contains_node(data.target_assignment.replicas, _self)) {
-            co_return co_await shutdown_on_current_shard(
-              std::move(ntp), command_rev);
-        }
-
-        /**
-         * in this case the partition is moved away from this node/shard, we may
-         * have to execute reconfiguration as this node may be a leader.
-         * Reconfiguration may involve triggering, cancelling or aborting
-         * configuration change.
-         *
-         */
-        auto ec = co_await execute_reconfiguration(
-          type, ntp, data, command_rev);
-        if (ec) {
-            co_return ec;
-        }
-        // Wait fo the operation to be finished on one of the nodes
-        co_return errc::waiting_for_reconfiguration_finish;
+    if (
+      partition
+      && (!expected_log_rev || partition->get_log_revision_id() < *expected_log_rev)) {
```

Comment:
do we have to check if an opeartion was finished and there are no in flight updates ?

Reply:
We don't have to - if this log_revision check yields true, this means that the existing partition belonged to a previous update epoch, or even a previous topic incarnation so we can safely delete it. If there is an in-flight update that needs this partition instance to succeed, `partition->get_log_revision_id()` would be equal to `*expected_log_rev`